src/site/xdoc/server/dev-extend.xml

```diff
@@ -72,6 +72,61 @@
   
   </subsection>
 
+    <subsection name="Extension and Guice wiring">
+
+        <p>This section does not concern the Spring / server/app wiring.</p>
+
+        <p>Guice applications let you load several type of user defined components. These components include:</p>
+
+        <ul>
+            <li><a href="https://github.com/apache/james-project/blob/master/mailet/api/src/main/java/org/apache/mailet/Mailet.java">Mailets</a></li>
+            <li><a href="https://github.com/apache/james-project/blob/master/mailet/api/src/main/java/org/apache/mailet/Matcher.java">Matchers</a></li>
+            <li><a href="https://github.com/apache/james-project/blob/master/mailbox/api/src/main/java/org/apache/james/mailbox/events/MailboxListener.java">Mailbox Listeners</a></li>
+            <li><a href="https://github.com/apache/james-project/blob/master/mailbox/api/src/main/java/org/apache/james/mailbox/extension/PreDeletionHook.java">PreDeletion hooks</a></li>
+            <li><a href="https://github.com/apache/james-project/blob/master/protocols/api/src/main/java/org/apache/james/protocols/api/handler/ProtocolHandler.java">Protocol handlers</a> (SMTP/LMTP/POP3)</li>
+            <li><a href="https://github.com/apache/james-project/blob/master/server/protocols/webadmin/webadmin-core/src/main/java/org/apache/james/webadmin/Routes.java">Additional webAdmin routes</a></li>
+        </ul>
+
+        <p>Given a custom maven module implementing one or more of the following APIs, first create a
+        <a href="https://maven.apache.org/plugins/maven-assembly-plugin/descriptor-refs.html#jar-with-dependencies">jar-with-dependencies</a>
+        of your module. Then copy the jar-with-dependencies into the <code>./extensions-jars</code> folder of your guice
+        James server installation.</p>
+
+        <p>Note: Transitive dependency resolution is only done if you are packaging your extension into a jar-with-dependencies.
+        Your extension class-loader will be the modified one (James and extensions-jars content) but James class-loader
+        stays unmodified everywhere else.</p>
+
+        <p>Once done, you need to explicitly require James to load your extensions. This needs to be done in the following
```

Comment:
Why don't we use an SPI for that? (https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html)
It's the standard way of implementing extensions so why would we create a custom solution for that?

Reply:
How does  SPI interacts with injection system?

Do you want the users extensions to fully rely on SPI for their dependency resolution?

I would consider this being a burdon...

Reply:
SPI is just a way to declare extension points so that people can put an extension on the classpath an get it loaded.

AFAIK, it doesn't mean that you can't use @Inject or any other DI system.

What it does mean is: 
* we define clearly with a standard how to extend James
* we rely on a standard jar META-INF to declare extensions

Reply:
I got with your link how as an extension writer I can declare the Services I implement,

I did not get how as an extension user how I select the Services implementations I want to use....

What if a Jat contains listenerA & listenerB & I only want listenerA?

Reply:
it should be in two different files then

Reply:
Also does this mean a breaking change for already provided extensions (mailboxListeners?)

Reply:
yes